kotlin/samples/hello-workflow-fragment/src/main/java/com/squareup/sample/helloworkflowfragment/HelloFragmentViewRunner.kt

```diff
@@ -18,34 +18,30 @@ package com.squareup.sample.helloworkflowfragment
 import android.annotation.SuppressLint
 import android.view.View
 import android.widget.TextView
-import com.squareup.coordinators.Coordinator
 import com.squareup.sample.helloworkflowfragment.HelloWorkflow.Rendering
 import com.squareup.workflow.ui.LayoutBinding
 import com.squareup.workflow.ui.ViewBinding
-import io.reactivex.Observable
-import io.reactivex.disposables.CompositeDisposable
+import com.squareup.workflow.ui.ViewRegistry
+import com.squareup.workflow.ui.ViewRunner
 
-@Suppress("EXPERIMENTAL_API_USAGE")
-class HelloFragmentCoordinator(
-  private val renderings: Observable<out Rendering>
-) : Coordinator() {
-  private val subs = CompositeDisposable()
+@Suppress("EXPERIMENTAL_API_USAGE", "EXPERIMENTAL_OVERRIDE")
+class HelloFragmentViewRunner : ViewRunner<Rendering> {
+  private lateinit var messageView: TextView
 
-  @SuppressLint("SetTextI18n")
-  override fun attach(view: View) {
-    val messageView = view.findViewById<TextView>(R.id.hello_message)
-
-    subs.add(renderings.subscribe { rendering ->
-      messageView.text = rendering.message + " Fragment!"
-      messageView.setOnClickListener { rendering.onClick(Unit) }
-    })
+  override fun bind(
+    view: View,
+    registry: ViewRegistry
+  ) {
+    messageView = view.findViewById(R.id.hello_message)
   }
 
-  override fun detach(view: View) {
-    subs.clear()
+  @SuppressLint("SetTextI18n")
+  override fun update(newValue: HelloWorkflow.Rendering) {
```

Comment:
Something I really like with the single attach(view) is that technically you can be much "purer" by not having to specify the lateinit vars as class members, and just assign the findViewByIds to local variables.

Reply:
Weren't you considering that trick Helios did where `update` shares the lexical scope of `bind`?

Reply:
@zach-klippenstein I'm not sure what your'e referring to, I'll ask him.

@Blisse Yeah. :/ While you can theoretically be purer, it still requires you to play games if you want to avoid doing a lot of redundant findById calls on every update, and last I looked those are really terrible. I liked how this approach nudges you into one set up / lots of updates.

What would you think of ViewRunner being an abstract class rather than an interface, with a constructor that requires a View? That would be a lot closer to what happens on iOS, actually, where they have a subclass of `ViewController` that you must extend. Something like:

```
abstract class ViewRunner<ViewT : View, ValueT>(protected val view: ViewT) {
  abstract fun display(value: ValueT)
}
```

Reply:
Taking this further, the `display` method could be from an interface, and the abstract class used only by `LayoutBinding`. That way we avoid putting awkward restrictions on code construction systems like Helios is working on.

```
interface ViewRunner<ValueT> {
  fun display(value: ValueT)
}
```

```
abstract class LayoutRunner<ViewT: View, ValueT>(
  protected val view: ViewT
) : ViewRunner<ValueT> {
}
```

(Also should stop forcing `ValueT` to extend `Any`.)

Reply:
Oh, has to extend Any if I want to use the type info the way I am. Fine with me.

Reply:
I mean this sort of thing:
```kotlin
bind { view ->
  val myView = view.findViewById(R.id.my_view)
  update { data ->
    myView.text = data.text
  }
}
```

Reply:
This is shaping up nicely (more functional and less brittle than I sketched), and it gets rid of all those anonymous runner classes.

```
class AuthorizingViewRunner(
  view: View,
  registry: ViewRegistry
) : LayoutRunner<AuthorizingScreen>(view, registry) {
  private val messageView: TextView = view.findViewById(R.id.authorizing_message)

  override fun showValue(value: AuthorizingScreen) {
    messageView.text = value.message
  }

  companion object : ViewBinding<AuthorizingScreen> by LayoutBinding.of(
      R.layout.authorizing_layout, ::AuthorizingViewRunner
  )
}
```

Reply:
Having to manually plumb the `ViewRegistry` through is unfortunate, seems like boilerplate we should get rid of somehow.

Reply:
Oh yeah, derp. Totally fixable.

Reply:
```
class HelloViewRunner(view: View) : LayoutRunner<Rendering>(view) {
  private val messageView: TextView = view.findViewById(R.id.hello_message)

  override fun showValue(value: Rendering) {
    messageView.text = value.message
    messageView.setOnClickListener { value.onClick(Unit) }
  }

  companion object : ViewBinding<Rendering> by binding(
      R.layout.hello_goodbye_layout, ::HelloViewRunner
  )
}
```

Reply:
At this point the base class `LayoutRunner` is kind of pointless, even though it's a nice blank to fill in. And the `ViewRunner` interface has already evaporated. Let me see if I can tighten this up some more.